http-common/src/request.rs

```diff
@@ -179,61 +150,80 @@ async fn make_call<TUri, TRequest>(
     method: http::Method,
     uri: TUri,
     headers: Option<&[(&str, &str)]>,
+    max_retries: u32,
     body: Option<&TRequest>,
 ) -> std::io::Result<(hyper::StatusCode, hyper::HeaderMap, hyper::body::Bytes)>
 where
     hyper::Uri: TryFrom<TUri>,
     <hyper::Uri as TryFrom<TUri>>::Error: Into<http::Error>,
     TRequest: serde::Serialize,
+    TUri: Clone,
 {
-    let mut req = hyper::Request::builder().method(method).uri(uri);
-
-    if let Some(headers) = headers {
-        if let Some(headers_map) = req.headers_mut() {
-            for (key, value) in headers {
-                headers_map.insert(
-                    headers::HeaderName::from_str(key)
-                        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
-                    headers::HeaderValue::from_str(value)
-                        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+    let mut retry_num = 0;
+    loop {
+        let mut req = hyper::Request::builder()
+            .method(method.clone())
+            .uri(uri.clone());
+
+        if let Some(headers) = headers {
+            if let Some(headers_map) = req.headers_mut() {
+                for (key, value) in headers {
+                    headers_map.insert(
+                        headers::HeaderName::from_str(key)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+                        headers::HeaderValue::from_str(value)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+                    );
+                }
+            }
+        }
+
+        // `req` is consumed by both branches, so this cannot be replaced with `Option::map_or_else`
+        //
+        // Ref: https://github.com/rust-lang/rust-clippy/issues/5822
+        #[allow(clippy::option_if_let_else)]
+        let req = if let Some(body) = body {
+            let body = serde_json::to_vec(body)
+                .expect("serializing request body to JSON cannot fail")
+                .into();
+
+            req.header(hyper::header::CONTENT_TYPE, "application/json")
+                .body(body)
+        } else {
+            req.body(Default::default())
+        };
+        let req = req.expect("cannot fail to create hyper request");
+
+        let response = client.request(req).await;
+        match response {
+            Ok(response) => {
+                let (
+                    http::response::Parts {
+                        status, headers, ..
+                    },
+                    body,
+                ) = response.into_parts();
+
+                let body = hyper::body::to_bytes(body)
+                    .await
+                    .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+
+                return Ok((status, headers, body));
+            }
+            Err(err) if err.is_closed() && retry_num < max_retries => {
```

Comment:
Are you sure? https://docs.rs/hyper/0.14.13/hyper/struct.Error.html#method.is_closed says it's for the sender (of the request body), not the underlying connection.

Reply:
Yep. I wrote code to send 11 (or more) concurrent requests to IS, and looked at the error code.

Console Output:
![image](https://user-images.githubusercontent.com/13950901/135698876-d907640d-6982-4d0b-be68-7a1f719660d6.png)

edged generating the error with 11 requests: 
![image](https://user-images.githubusercontent.com/13950901/135698837-c317990a-7bb9-4d63-9809-6469837c6e69.png)

request.rs printing the error:
![image](https://user-images.githubusercontent.com/13950901/135698855-6ff75c68-5307-48ce-bb8a-df8489c5ce54.png)

Reply:
Remind me again - what does IS return when it throttles a request?

Reply:
The `Http Error: channel closed` in the logs above is from IS. The code I used to test just sends 11 concurrent requests and prints the error IS causes. 

IS isn't actually sending an error, we don't get far enough into the request handling to actually send one. IS just closes the incoming connection instead.